### PR TITLE
ARM64-SVE: Add `AddSequentialAcross`

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Follow-up to #104258, part of #99957. I'll do `AddRotateComplex` in a separate PR, since I think we might need to change the signature of that API if we want the user to pass the rotation constant in degrees: The type of the constant is currently `byte`, which is too small to handle a rotation amount of 270.

Test output:
```
Starting test: .\Core_Root\corerun.exe .\HardwareIntrinsics_Arm_r\HardwareIntrinsics_Arm_r.dll Sve_AddSequentialAcross
===================Running default===================
------------------- {} -------------------
Passed test: _Sve_r::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_AddSequentialAcross_float() : 31
Passed test: _Sve_r::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_AddSequentialAcross_double() : 31
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------

Starting test: .\Core_Root\corerun.exe .\HardwareIntrinsics_Arm_ro\HardwareIntrinsics_Arm_ro.dll Sve_AddSequentialAcross
===================Running default===================
------------------- {} -------------------
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_AddSequentialAcross_float() : 31
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_AddSequentialAcross_double() : 31
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------
```

@dotnet/arm64-contrib PTAL, thanks!

### CODE REVIEW

**Comment by kunalspathak** [comment]
> I'll do AddRotateComplex in a separate PR, since I think we might need to change the signature of that API if we want the user to pass the rotation constant in degrees: The type of the constant is currently byte, which is too small to handle a rotation amount of 270.

Please take up this discussion in the original APi review issue and tag the reviewers. @tannergooding

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs [hunk 1]
```diff
@@ -4136,6 +4136,32 @@ public static byte AddSaturate(byte op1, byte op2)
             return ovf ? byte.MaxValue : result;
         }
 
+        public static double[] AddSequentialAcross(double[] op1, double[] op2)
+        {
+            double[] result = new double[op1.Length];
+            result[0] = op1[0];
+
+            foreach (double num in op2)
```
Comment (kunalspathak) [inline src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs:4144]: This only calculates the sum for active lanes, so you need to check that as well.

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs [hunk 2]
```diff
@@ -3121,6 +3121,9 @@
...
 
+    ("SveVecBinOpVecTest.template",       new Dictionary<string, string> { ["TestName"] = "Sve_AddSequentialAcross_float",                                                                         ["Isa"] = "Sve",           ["LoadIsa"] = "Sve",     ["Method"] = "AddSequentialAcross",                                                  ["RetVectorType"] = "Vector",    ["RetBaseType"] = "Single",  ["Op1VectorType"] = "Vector",    ["Op1BaseType"] = "Single", ["Op2VectorType"] = "Vector",    ["Op2BaseType"] = "Single",                                                              ["LargestVectorSize"] = "64", ["NextValueOp1"] = "TestLibrary.Generator.GetSingle()",        ["NextValueOp2"] = "TestLibrary.Generator.GetSingle()",                                                                                        ["ValidateVectorResult"] = "!result.SequenceEqual(Helpers.AddSequentialAcross(left, right))",   ["GetVectorResult"] = "Helpers.AddSequentialAcross(left, right)"}),
```
Comment (kunalspathak) [inline src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:3124]: does this test the operation with `ConditionalSelect` as well?

Comment (amanasifkhalid) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:3124]: It does, using `SimpleVecOpTest_VectorValidationLogicForCndSel` and `SimpleVecOpTest_VectorValidationLogicForCndSel_FalseValue`. I guess I've been getting lucky with `mask[0]` being 0 for the `CndSel` scenarios that skip lanes in the second op.

File: src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs (hunk 1 above)
Comment (amanasifkhalid) [reply src/tests/JIT/HardwareIntrinsics/Arm/Shared/Helpers.cs]: Got it, fixed.

**Comment by amanasifkhalid** [comment]
@kunalspathak thanks for the review -- I've fixed the helpers to skip inactive lanes. Stress tests are still passing.

**Comment by kunalspathak** [comment]
> I've fixed the helpers to skip inactive lanes.  Stress tests are still passing.

If the tests were passing without fixing for inactive lanes, that means the `mask` input given was always making all lanes active. Can you please double check if the `mask` input produced is generating `0` and `1` both? There are helpers like `Helpers.getMask*()` to produce them.

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs (hunk 2 above)
Comment (kunalspathak) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:3124]: We should use `Helpers.getMask*()` to produce the `mask` input.

Comment (amanasifkhalid) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:3124]: Got it, I tweaked the test definitions to use `getMask*` for `NextValueOp1`. The only odd thing about this approach is the template for `SveVecBinOpVecTest` uses `NextValueOp1` for generating the mask input, and for `op1`'s input, so the test values for `op1` are artificially constrained by this approach. I can tweak the test template to use separate template variables for these two, but lots of existing tests would need to be updated, and I don't think the value gained from doing this is much.

Comment (kunalspathak) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:3124]: I am fine with not doing it in this PR, but can you please open an issue to update the templates to have them separate?

Comment (amanasifkhalid) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:3124]: Sure thing: https://github.com/dotnet/runtime/issues/104804

**Review by kunalspathak — APPROVED** [review]
LGTM. Thanks!

**Comment by kunalspathak** [comment]
It occured to me that this instruction produces scalar results, so if you combine it with `ConditionalSelect`, which produces results in scalable register, you might be testing the results in all lanes, which will be wrong. You might just need to verify the results in lane 0.

```
fadda v9, p0, v1, z2.d
sel z8, p0, z9, z7
```

Above, the results are only present in lane 0 of `v9` / `z9`. So you might want to verify just for that lane. All the other lanes will hold the value that was there before performing `fadda`.

**Comment by amanasifkhalid** [comment]
After discussing offline with @kunalspathak, we've decided to match `AddAcross` for now and skip the `CndSel` scenarios for this API, as there's some ambiguity as to what the shape of these scenarios should look like. See #101770, and #101974 for updating these APIs with explicitly-masked variants.

**Comment by amanasifkhalid** [comment]
I've removed the CndSel scenarios for now. Stress tests are passing.

**Comment by amanasifkhalid** [comment]
@kunalspathak are you ok with me merging this as-is?

**Review by kunalspathak — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**